code/game/gamemodes/meteor/meteors.dm

```diff
@@ -90,56 +89,81 @@ GLOBAL_LIST_INIT(meteorsC, list(/obj/effect/meteor/dust)) //for space dust event
 	icon_state = "small"
 	density = TRUE
 	anchored = TRUE
-	var/hits = 4
-	var/hitpwr = 2 //Level of ex_act to be called on hit.
-	var/dest
 	pass_flags = PASSTABLE
+
+	///The resilience of our meteor
+	var/hits = 4
+	///Level of ex_act to be called on hit.
+	var/hitpwr = 2
+	//Should we shake people's screens on impact
 	var/heavy = FALSE
+	///Sound to play when you hit something
 	var/meteorsound = 'sound/effects/meteorimpact.ogg'
+	///Our starting z level, prevents infinite meteors
 	var/z_original
-	var/threat = 0 // used for determining which meteors are most interesting
-	var/lifetime = DEFAULT_METEOR_LIFETIME
-	var/timerid = null
+	///Used for determining which meteors are most interesting
+	var/threat = 0
+
+	//Potential items to spawn when you die
 	var/list/meteordrop = list(/obj/item/stack/ore/iron)
+	///How much stuff to spawn when you die
 	var/dropamt = 2
 
-/obj/effect/meteor/Move()
-	if(z != z_original || loc == dest)
-		qdel(src)
-		return FALSE
+	///The thing we're moving towards, usually a turf
+	var/atom/dest
+	///Lifetime in seconds
+	var/lifetime = DEFAULT_METEOR_LIFETIME
+
+/obj/effect/meteor/Initialize(mapload, turf/target)
+	. = ..()
+	z_original = z
+	GLOB.meteor_list += src
+	SSaugury.register_doom(src, threat)
+	SpinAnimation()
+	chase_target(target)
+
+/obj/effect/meteor/Destroy()
+	GLOB.meteor_list -= src
+	SSaugury.unregister_doom(src)
+	. = ..()
+
+/obj/effect/meteor/Moved(atom/OldLoc, Dir, Forced = FALSE)
+	. = ..()
 
-	. = ..() //process movement...
+	var/dest_turf = get_turf(dest) //Dest can be deleted in the if block
 
-	if(.)//.. if did move, ram the turf we get in
+	if(OldLoc != loc)//If did move, ram the turf we get in
 		var/turf/T = get_turf(loc)
 		ram_turf(T)
 
 		if(prob(10) && !isspaceturf(T))//randomly takes a 'hit' from ramming
 			get_hit()
 
-/obj/effect/meteor/Destroy()
-	if (timerid)
-		deltimer(timerid)
-	GLOB.meteor_list -= src
-	SSaugury.unregister_doom(src)
-	walk(src,0) //this cancels the walk_towards() proc
-	. = ..()
+	if(z != z_original || loc == get_turf(dest_turf))
+		qdel(src)
+		return
 
-/obj/effect/meteor/Initialize(mapload, target)
-	. = ..()
-	z_original = z
-	GLOB.meteor_list += src
-	SSaugury.register_doom(src, threat)
-	SpinAnimation()
-	timerid = QDEL_IN(src, lifetime)
-	chase_target(target)
+/obj/effect/meteor/Process_Spacemove()
+	return TRUE //Keeps us from drifting for no reason
 
 /obj/effect/meteor/Bump(atom/A)
+	. = ..() //What could go wrong
 	if(A)
 		ram_turf(get_turf(A))
 		playsound(src.loc, meteorsound, 40, TRUE)
 		get_hit()
 
+/obj/effect/meteor/proc/chase_target(atom/chasing, delay, home)
+	if(!isatom(chasing))
+		return
+	SSmovement_loop.move_towards(src, chasing, delay, home, lifetime)
+	RegisterSignal(src, COMSIG_MOVELOOP_END, .proc/handle_stopping)
+
+/obj/effect/meteor/proc/handle_stopping()
+	SIGNAL_HANDLER
+	if(!QDELETED(src))
```

Comment:
```suggestion
	if(!QDELING(src))
```

Reply:
`COMSIG_PARENT_QDELETING` causes the movement datum to remove itself from the subsystem, which in turn calls `kill()` on the movement datum. `kill()` sends `COMSIG_MOVELOOP_END`, and the cycle repeats. The solution for this I think is to move the `QDELETED()` check up a level to `kill()` itself, but that does mean that something that uses `COMSIG_MOVELOOP_END` to modify an object aside from itself will have odd behavior. 

It's quite a mess, do you have any advice?

Reply:
> The solution for this I think is to move the `QDELETED()` check up a level to `kill()` itself, but that does mean that something that uses `COMSIG_MOVELOOP_END` to modify an object aside from itself will have odd behavior.

Restructure your code so that it doesn't loop around.
Centralize the responsibility in the subsystem, so that all relevant changes go through it.
Or on the datum.
Take your pick, but don't have messy looping logics, with distributed responsibilities that are hard to track.